cpp/arcticdb/vector_db/vector_db.cpp

```diff
@@ -0,0 +1,102 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/processing/clause.hpp>
+#include <arcticdb/vector_db/vector_db.hpp>
+
+namespace arcticdb {
+    TopKClause::TopKClause(std::vector<double> query_vector, uint8_t k) :
+            query_vector_(std::move(query_vector)),
+            k_(k) {
+        clause_info_.modifies_output_descriptor_ = true;
+        clause_info_.follows_original_columns_order_ = false;
+        clause_info_.can_combine_with_column_selection_ = false;
+    }
+
+    Composite<ProcessingSegment> TopKClause::process(std::shared_ptr<Store> store,
+                                                     Composite<ProcessingSegment> &&p) const {
+        auto procs = std::move(p);
+        TopK top_k(k_);
+        // We expect a vector in each column.
+        // top_k_ is updated as we read more vectors. We want to efficiently remove
+        // the 'worst' (least similar) vector as we go along to ensure we have the
+        // top k, not the top k+1.
+        auto lp = 2;
+        // Let's pretend that only the lp-norms exist. We'll use p=2 for now as a default.
+
+        procs.broadcast(
+                [&store, &top_k, lp, this](const ProcessingSegment &proc) {
+                    for (const auto& slice_and_key: proc.data()) {
+                        std::vector <std::shared_ptr<Column>> columns = slice_and_key.segment(store).columns();
+                        for (auto&& [idx, col]: folly::enumerate(columns)) {
+                            internal::check<ErrorCode::E_ASSERTION_FAILURE>(
+                                    static_cast<long unsigned>(col->row_count()) == this->query_vector_.size(),
+                                    "Expected vector of length {}, got vector of length {}.",
+                                    col->row_count(),
+                                    this->query_vector_.size());
+                            col->type().visit_tag([&top_k, &slice_and_key, &store, &col=col, idx=idx, lp, this](auto type_desc_tag) {
+                                using TypeDescriptorTag = decltype(type_desc_tag);
+                                using RawType = typename TypeDescriptorTag::DataTypeTag::raw_type;
+
+                                if constexpr (is_floating_point_type(TypeDescriptorTag::DataTypeTag::data_type)) {
```

Comment:
Add an `else{internal::raise...}`, otherwise it will silently do nothing/fail later with non-float columns